core/src/main/java/com/orientechnologies/orient/core/storage/cache/local/OWOWCache.java

```diff
@@ -667,60 +659,166 @@ public Future store(final long fileId, final long pageIndex, final OCachePointer
 
   public OCachePointer[] load(long fileId, long startPageIndex, int pageCount, boolean addNewPages, OModifiableBoolean cacheHit,
       boolean verifyChecksums) throws IOException {
+
     final int intId = extractFileId(fileId);
     if (pageCount < 1)
       throw new IllegalArgumentException("Amount of pages to load should be not less than 1 but provided value is " + pageCount);
 
     filesLock.acquireReadLock();
     try {
-      final PageKey[] pageKeys = new PageKey[pageCount];
-      for (int i = 0; i < pageCount; i++) {
-        pageKeys[i] = new PageKey(intId, startPageIndex + i);
-      }
+      //first check that requested page is already cached so we do not need to load it from file
+      final PageKey startPageKey = new PageKey(intId, startPageIndex);
+      final Lock startPageLock = lockManager.acquireSharedLock(startPageKey);
+
+      //check if page already presented in write cache
+      final PageGroup startPageGroup = writeCachePages.get(startPageKey);
+
+      //page is not cached load it from file
+      if (startPageGroup == null) {
+        //load it from file and preload requested pages
+        //there is small optimization
+        //if we need single page no need to release already locked page
+        Lock[] pageLocks;
+        PageKey[] pageKeys;
+
+        if (pageCount > 1) {
+          startPageLock.unlock();
+
+          pageKeys = new PageKey[pageCount];
+          for (int i = 0; i < pageCount; i++) {
+            pageKeys[i] = new PageKey(intId, startPageIndex + i);
+          }
 
-      Lock[] pageLocks = lockManager.acquireSharedLocksInBatch(pageKeys);
-      try {
-        PageGroup pageGroup = writeCachePages.get(pageKeys[0]);
+          pageLocks = lockManager.acquireSharedLocksInBatch(pageKeys);
+        } else {
+          pageLocks = new Lock[] { startPageLock };
+          pageKeys = new PageKey[] { startPageKey };
+        }
 
-        if (pageGroup == null) {
-          final OCachePointer pagePointers[] = cacheFileContent(intId, startPageIndex, pageCount, addNewPages, cacheHit,
-              verifyChecksums);
+        OCachePointer pagePointers[];
+        try {
+          //load requested page and preload requested amount of pages
+          pagePointers = loadFileContent(intId, startPageIndex, pageCount, verifyChecksums);
 
-          if (pagePointers.length == 0)
-            return pagePointers;
+          if (pagePointers != null) {
+            if (pagePointers.length == 0)
```

Comment:
Yep, I know that is implemented in this way )) but IMHO we should not keep dead code it makes understanding of code worse than it could be.